### ISSUE TITLE
vgrow and hgrow helper functions

### ISSUE THREAD

**Issue by voddan** [edvin/tornadofx#204]
I found myself using those 2 helper functions quite a lot:

    fun Node.hgrow(priority: Priority) = HBox.setHgrow(this, priority)
    fun Node.vgrow(priority: Priority) = VBox.setVgrow(this, priority)

Apart from being short, those functions are well discoverable in the code completion. This is despite the fact that JavaFX has a very unusual (mind-bending) concept setting growth. 

I guess it also can be implemented as properties.

**Comment by thomasnield** [comment]
This is within an `HBox` and `VBox` context though. I don't know if those extension function names make that clear.

**Comment by voddan** [comment]
Yes, they have effect only inside the boxes. Altho, calling them somewhere else doe no harm.

**Comment by thomasnield** [comment]
It does not do any harm, but I think it can still cause confusion when people thinks it will work outside an `HBox` or `VBox`. We do have the `hboxconstraints()` and `vboxconstraints()` blocks for any `Node`. This helps keep the context somewhat clear. 

```kotlin
class MyView : View("My View") {
    override val root = vbox {
        button("Press me") {
            hboxConstraints {
                hGrow = Priority.ALWAYS
            }
        }
    }
}
```

**Comment by voddan** [comment]
Technically you are correct. But practically  `hboxConstraints` is mostly useless. 

Because:

- If I am not familiar with it, then I search `this@node.grow<something>` and do not find  `hboxConstraints`. Even if I do see it, its visual semantics is "creating something", not "setting something (like `hgrowPriority`)".

- If I do know about `hboxConstraints`, then it is 17 (!!) characters longer than `hgrow(...)`

Bottom line the current API is not satisfying because of (a) inherited corks of JavaFX and (b) syntax which does not support search by code completion.

**Comment by voddan** [comment]
I (as a user) prefer the look of my api (at the top of this issue), and maybe it is okey if I get runtime exceptions when I misuse it.

**Comment by thomasnield** [comment]
Okay I guess I agree with the need to shorthand. But there's got to be silver bullet to this... maybe the function can invoke `hgrow`/`vgrow` for all possible layouts?

**Comment by edvin** [comment]
That's a good idea Thomas. If hgrow / vgrow makes sense in a layout we know about, we can make sure to set the correct properties. In any case, I don't have any issues with adding these shortcuts, since even the original JavaFX API suffers from that you can set these properties on any Node even if they are not surrounded by an H/VBox. Therefore I don't think we would add to the confusion if we support hgrow/vgrow directly on Node. We can start with just setting `HBox.setHgrow` or `VBox.setVgrow` and then grow them as we find other layouts where these semantics make sense?

**Comment by thomasnield** [comment]
That sounds good!

**Comment by edvin** [comment]
Ok, so I added them as getters/setters like this:

```kotlin
var Node.hgrow: Priority get() = HBox.getHgrow(this); set(value) { HBox.setHgrow(this, value) }
var Node.vgrow: Priority get() = VBox.getVgrow(this); set(value) { VBox.setVgrow(this, value) }
```

**Comment by voddan** [comment]
That was quick! Thanks!

**Comment by edvin** [comment]
@voddan No problem :) By the way, I did not add them as functions, I wanted them to behave like properties so you can actually query for the value as well. Hope that's OK.

**Comment by voddan** [comment]
Even better! Thank you! 

Btw, off-topic: I am now investigating a strange memory leak out of JavaFX bindings. Do you know any good/simple guide for troubleshooting those things?

**Comment by edvin** [comment]
I don't have much experience with that, but maybe you could ask in the Slack channel?

**Comment by Hillrunner2008** [comment]
@voddan, you may want to take a look at https://bugs.openjdk.java.net/browse/JDK-8089305. Also it is worth mentioning that it is important to use the weak reference version of the various listeners in many cases to prevent memory leaks. Hope this helps in some way.
